[클린코드 리액트 2기 장태영] 장바구니 미션 Step 2

taeyoungs opened the pull request:
# Pull Request Template

## 1. 가이드

### 1-1. 배포

- [Storybook](https://6415ad1752206cf729a787bc-allgqvfali.chromatic.com/?path=/story/common-button--template&args=type:primary)

### 1-2. 로컬 환경

```bash
cd client

# install
yarn

# run
yarn start
```

## 2. 리뷰 요청 ✍️

중간에 체크된 항목들을 관리하는 방식을 바꾸었습니다. 관련 코드(테스트 코드 포함)가 전부 변경되는 부분이 있어 혹시나 리뷰하실 때 보기 불편하실 수도 있을 것 같아 미리 코멘트를 남겨 놓습니다. 😭

### 2-1. 어렵거나 아쉬웠던 점

- 상태 관리 라이브러리로 `zustand`를 선택했는데 처음 사용해보다 보니 '이렇게 하는게 맞나..?' 라는 생각이 드는 부분이 있었습니다. 사용법은 다음의 링크들을 참조했습니다.
  - https://docs.pmnd.rs/zustand/getting-started/introduction
  - https://tkdodo.eu/blog/working-with-zustand
  - `zustand`를 선택한 이유는 'Redux와 사용법이나 데이터 흐름이 비슷하면서 더 적은 코드로 간단하고 빠르게 스토어를 만들 수 있는 것 같아서' 였습니다. 테스트 전략을 세울 때 Redux를 참고할 수도 있지 않을까 싶기도 했구요. 
- '페이징 혹은 인피니티 스크롤 적용'은 아직 구현하지 못한 상태인데 현재 PR 올린 후에 진행해보도록 하겠습니다.

### 2-2. 나누고 싶은 고민

- [Redux에서 reducer에 대해 테스트](https://redux.js.org/usage/writing-tests#reducers)를 하듯이 zustand로 만든 스토어의 action 내에서 사용되는 로직을 순수함수로 만들고 해당 함수에 대한 테스트 코드(cartModel.test.ts)를 작성했습니다. 여기다가 추가적으로 action을 호출하기 전/후 스토어의 값에 대한 테스트 코드(cart.test.ts)를 작성하긴 했는데 .. 전체적으로 어색하고 구멍이 뚫려있는 느낌입니다. 🥲 
보시고 걸리시는 부분이 있다면 알려주셨으면 좋겠습니다! 🙇‍♂️ (ex. 이 부분은 불필요한 것 같다, 이런 전략이나 방향성은 어떤가 등)

### 2-3. 중점적으로 리뷰받고 싶은 부분

### 2-4. 새로운 시도 혹은 미션을 통해 도전한 점

- 비교적 익숙한 Redux 대신 zustand를 상태 관리 라이브러리로 택했습니다.
- 평소에는 목록에서 체크 여부를 판단할 때 isChecked라는 속성을 따로 추가해주곤 했는데 이 방법 대신 Set 자료구조를 이용해서 체크된 항목들의 id를 관리하는 방식으로 진행해보았습니다.

### 2-5. 도식화된 자료 혹은 작성된 이미지 형태의 설계 구조

## 3. 질문있어요 🙋

- 질문이라기엔 애매하지만 '나누고 싶은 고민'란에 적어두었습니다.

## 4. 요구 사항 ✅

### 필수 요구사항

- [x]  MSW를 활용한 `API mocking`
- [x]  `Endpoint`만 변경하면 언제든 `Real API`를 바라볼 수 있다고 가정하고 상상한다.
    - [x]  `Real API` 없이 로컬에서만 동작하는 상태로 리뷰 받는 것이 기본 원칙
- [x]  상태 관리 라이브러리를 선택하고 적용한다.
    - [x]  전략을 세우고 PR 본문에 내용을 작성한다.
- [x]  선택한 상태 관리 라이브러리에 대응되는 테스트 전략을 세우고 PR 본문에 내용을 작성한다.
    - [x]  없다면 `React Testing Library` & `Jest`를 활용해 자유로운 단위의 테스트라도 진행한다.

### 상품목록

- [ ]  페이징 혹은 인피니티 스크롤 적용
    - [ ]  뒤로가기 및 페이지 전환시 기존 페이지 및 스크롤 위치 기억

### 장바구니

- [x]  해당 상품의 수량을 변경할 수 있다.
    - [x]  상품의 수량은 항상 1이상, 20이하여야 한다
        - [x]  상품의 수량이 1이면 상품 수량 감소할 수 없다.
        - [x]  상품의 수량이 20이면 상품 수량 증가할 수 없다.
    - [x]  해당 상품의 총 금액이 변경된다.
    - [x]  해당 상품이 체크되어있으면, 결제예상금액도 변경된다.
- [x]  모두선택 버튼이 체크되면, 상품들이 모두 선택된다.
    - [x]  모두선택 버튼이 체크가 풀리면, 상품들의 선택이 모두 해제된다.
- [x]  상품 삭제 버튼을 누르면, confirm 메시지가 보여진다.
    - [x]  확인을 누르면, 선택된 상품이 모두 삭제된다.
    - [x]  결제예상금액이 0원이 된다.
- [x]  🗑 버튼을 누르면 confirm 메시지가 보여진다.
    - [x]  확인을 누르면, 해당 상품이 삭제된다.
- [x]  체크된 상품 개수에 따라 주문하기 버튼 내부에 수량이 변경된다.
- [x]  주문하기 버튼을 누르면, confirm 메시지가 보여진다.
    - [x]  확인을 누르면, 주문/결제 페이지로 이동한다.
    - [x]  확인을 누르면, 장바구니에서 선택된 상품들이 삭제된다.
    - [x]  확인을 누르면, 체크된 상품들을 데이터베이스에서 제거한다.
- [x]  주문할 상품이 0개이면 버튼이 비활성화된다.

### 4-2. 선택 요구사항 (심화)

### 상품상세

- [x]  페이지에는 상품 사진, 이름, 금액 정보가 보여진다.
- [x]  장바구니 버튼을 클릭하면 장바구니 페이지로 이동한다.
- [x]  장바구니 버튼을 클릭하면 해당 상품이 장바구니에 담긴다.

### UX/UI

- [ ] 사용성 개선
  - [x] 사용자를 위한 로딩 환경 개선

taeyoungs committed:
chore: 상태 관리 라이브러리 zustand 추가

taeyoungs committed:
mod: useFetch에 onSuccess, onError props 추가

taeyoungs committed:
chore: handler 파일을 cart와 product 파일로 분리

taeyoungs committed:
feat: cart store에서 사용할 함수 및 테스트 추가

taeyoungs committed:
feat: Cart store 및 store action 테스트 코드 추가

taeyoungs committed:
feat: 결제예상금액 섹션을 컴포넌트로 분리

taeyoungs committed:
mod: cart store를 이용한 방식으로 CartItem 업데이트

taeyoungs committed:
feat: cart store를 이용한 방식으로 CartList 업데이트

taeyoungs committed:
mod: API Endpoint 변경 및 삭제 API body 수정

mocks/handlers/cart.ts
- Cart 객체가 담긴 배열 대신 Cart 객체의 id가 담긴 배열을
  Request Body로 받게끔 수정

taeyoungs committed:
mod: useFetch에 refetch용 trigger 변수 및 함수 추가

taeyoungs committed:
feat: 장바구니 삭제를 위한 mutation 함수 추가

taeyoungs committed:
feat: 체크된 항목들의 id 값을 얻어내는 함수 및 훅

taeyoungs committed:
feat: 장바구니 항목 삭제 mutation 래핑 훅 추가

taeyoungs committed:
feat: 장바구니 항목 삭제 기능 구현

taeyoungs committed:
chore: orders API 경로 추가 및 carts API 경로 수정

taeyoungs committed:
feat: 주문하기에 대한 Mock API(addOrder) 추가

taeyoungs committed:
feat: useRouter 커스텀 훅 구현

hooks/useRouter.ts
- PATHS 상수에 담긴 경로들이 아닌 경로를 인수로 넘기려고 할 때
  타입 에러가 발생할 수 있게 만들기 위해 useNavigate의 navigate
  함수를 래핑한 커스텀 훅 추가

taeyoungs committed:
feat: Order 객체에 대한 스키마 및 타입 추가

taeyoungs committed:
feat: 주문하기 API를 위한 mutation 함수 추가

taeyoungs committed:
feat: 장바구니 페이지 주문하기 기능 구현

taeyoungs committed:
feat: 주문/결제 페이지 추가

taeyoungs committed:
test: id로 항목을 제거하는 로직에 대한 테스트 추가

taeyoungs committed:
feat: ErrorBoundary 컴포넌트 추가

taeyoungs committed:
chore: 훅 네이밍 변경 (useCarts => useCartList)

taeyoungs committed:
mod: 체크된 항목 관리를 Set을 이용한 방식으로 변경

taeyoungs committed:
mod: 장바구니 store 변경사항 반영

taeyoungs committed:
feat: ErrorBoundary DefaultFallback 컴포넌트 추가

taeyoungs committed:
mod: Suspense를 ErrorBoundary로 래핑

taeyoungs committed:
chore: 주석 및 불필요한 함수 호출 제거

taeyoungs committed:
feat: 장바구니에 담긴 상품이 없을 때의 문구 추가

taeyoungs committed:
test: toBe에서 toEqual로 수정

jho2301 commented on client/src/api/deleteCarts.ts:
```diff
@@ -0,0 +1,17 @@
+  const data = await response.json();
+
+  return CartsSchema.parse(data);
+}
```
5-14 라인은 사용하는 api 에 무관하게 추상화해볼 수 있을 것으로 보입니다 
개인적으로는 apiClient.delete 의 형태로 메서드를 호출할 수 있으면 좋을 것 같네요!

taeyoungs replied:
나름대로 적용을 해보았는데 제가 잘 이해하고 적용한건지 모르겠네요. 😇
[adf0b28](https://github.com/next-step/react-shopping-cart/pull/23/commits/adf0b2828e1d581b83ff6170a78c705d074c59d3)

jho2301 commented on client/src/components/ErrorBoundary/index.tsx:
```diff
@@ -0,0 +1,38 @@
+  hasError: boolean;
+}
+
+// https://ko.reactjs.org/docs/error-boundaries.html
```
소소 
```js
/** @see https://ko.reactjs.org/docs/error-boundaries.html */
```
jsdoc형태로 사용하면 ide에서 바로 페이지로 진입할 수 있어서 보다 좋은 것 같아요

taeyoungs replied:
간단하지만 좋은 방법이네요. 👍 
[a1a0ce0](https://github.com/next-step/react-shopping-cart/pull/23/commits/a1a0ce0bdc58100b7afcda1a23bfcf9a3380e87d)

jho2301 commented on client/src/hooks/useFetch.ts:
```diff
@@ -48,13 +56,20 @@ function useFetch<T = unknown>({ fetcher, cacheKey, cacheTime }: UseFetchProps<T
+
+  function refetch() {
+    setTrigger((prevState) => !prevState);
+  }
```
shouldRefetch 정도로 이름에서 역할을 구체화시켜봐도 좋을 것 같아요

taeyoungs replied:
[bfcbf39](https://github.com/next-step/react-shopping-cart/pull/23/commits/bfcbf390af07fe37a8cec9e9e565ea9fce8b64f6)

jho2301 commented on client/src/pages/Cart/components/CartItem/index.tsx:
```diff
@@ -1,21 +1,56 @@
+  const { count: defaultCount, product, id } = cart;
   const { imageUrl, name, price } = product;
   const { count, minus, plus } = useCounter(defaultCount);
+  const { mutate: deleteCarts, isLoading } = useDeleteCarts({ onSuccess: refetchCarts });
```
useQuery의 mutateAsync 와 같이 훅에 콜백을 전달하지 않고 프로미스형태로 deleteCarts를 사용할 수 있게되면,   `handleClickMinusButton` 에서 핸들링로직 일체가 표현되어서 보다 동작을 명확하게 파악할 수 있는 핸들러함수가 될 것 같다는 생각이 있습니다

taeyoungs replied:
간단할 줄 알았는데 이래저래 시행착오를 조금 겪었습니다. 우선 의도대로 동작하긴 합니다.. 🥲
[4df9a78](https://github.com/next-step/react-shopping-cart/pull/23/commits/4df9a784fac9ae5ff989636e14d29d844ce63619)

jho2301 commented on client/src/pages/Cart/components/CartItem/index.tsx:
```diff
@@ -1,21 +1,56 @@
+    toggle: toggleIsChecked,
+    increase: increaseCartById,
+    decrease: decreaseCartById,
+  } = useCartActions();
```
```suggestion
  const cartAction = useCartActions();
  
  cartAction.toggle()
```

취향의 영역인 것 같긴하지만, rename할때  rename이 같이 적용이 안돼서 위랑 같이 사용하는 것을 선호합니다. 객체가 동작의 주체가 되어서 보다 명시적이라고 생각했어요

taeyoungs replied:
> 객체가 동작의 주체가 되어서 보다 명시적이라고 생각했어요

구조 분해 할당을 할 때마다 명시적이지 않다고 생각해서 rename을 하고 있었습니다. 그렇다고 useCartActions에서 반환해주는 함수의 네이밍을 위에서 제가 rename 한 것 처럼 바꾸기엔 그것도 뭔가 이미 useCartActions라는 훅의 이름에서 주체를 명시하고 있어서 그런지 불필요하게 두 번 명시하는 듯한 느낌을 받았거든요.

생각이 갇혀 있었던 것 같네요. 감사합니다.
이것도 추천해주신 방법이 마음에 들어서 적용해보도록 하겠습니다!

taeyoungs replied:
[b4a140d](https://github.com/next-step/react-shopping-cart/pull/23/commits/b4a140de3a7f0f92d57fb35c26677995259d33b6)

jho2301 commented on client/src/pages/Cart/components/CartItem/index.tsx:
```diff
@@ -1,21 +1,56 @@
+  };
+
+  const handleClickPlusButton = () => {
+    plus();
```
요것도 객체분해하지 않고 counter 객체를 사용했다면  counter.plus로 보다 하는 일이 명확한 함수가 됐을 것 같아요

taeyoungs replied:
> 요것도 객체분해하지 않고 counter 객체를 사용했다면 counter.plus로 보다 하는 일이 명확한 함수가 됐을 것 같아요

이렇게만 해도 확실히 주체가 표현되네요! 감사합니다. 훨씬 좋습니다. 😁

jho2301 commented on client/src/pages/Cart/index.tsx:
```diff
@@ -8,9 +11,11 @@ function Cart() {
+        <Suspense fallback={<CartListFallback />}>
+          <CartList />
+        </Suspense>
+      </ErrorBoundary>
```
페이지가 미처 핸들링하지 못한 에러도 처리하는 전역 에러바운더리가 있으면 좋을 것 같아요

taeyoungs replied:
[58330ab](https://github.com/next-step/react-shopping-cart/pull/23/commits/58330abdc1816ed87c6adbddfd275c1655b1ff93)

jho2301 commented on client/src/store/cart/index.ts:
```diff
@@ -0,0 +1,76 @@
+    const newCheckedIds = toggleCart(checked, id);
+
+    useCartStore.setState(() => ({ checked: newCheckedIds }));
+  };
```
요거 이전에 달았던 코멘트인지 헷갈리네요. 커스텀훅은 어디서 사용될지에 대한 정보를 알지 못하기 때문에 최대한 메모이제이션이 되면 좋을 것 같아요!

taeyoungs replied:
> 요거 이전에 달았던 코멘트인지 헷갈리네요. 커스텀훅은 어디서 사용될지에 대한 정보를 알지 못하기 때문에 최대한 메모이제이션이 되면 좋을 것 같아요!

맞습니다! 이전에도 해주셨던 코멘트입니다. 제가 또 깜박했네요. 반영해놓도록 하겠습니다. 🙇‍♂️

taeyoungs replied:
[6d7e237](https://github.com/next-step/react-shopping-cart/pull/23/commits/6d7e237c705437618bbf9d0e04fc7acd35ba4d14)

jho2301 commented on client/src/store/cart/index.ts:
```diff
@@ -0,0 +1,76 @@
+
+export const useCartStore = create<State & Action>()((set) => ({
+  carts: [],
+  checked: new Set(),
```
변경사항에 대응할 때 결국엔 체크된 아이템에 대한 정보를 사용하게 되는경우가 필연적으로 있어서, 아이템 객체를 들고있어도 좋지 않을까 생각해봅니다. 

현재의 경우에도 
```ts
export const totalPriceOfCheckedCarts = (carts: Carts, checkedIds: Set<number>) => {
  return carts.reduce((total, { id, product: { price }, count }) => {
    if (checkedIds.has(id)) {
      return total + price * count;
    } else {
      return total;
    }
  }, 0);
};
```
아이템 객체리스트를 들고있었다면 보다 로직이 간결해졌을 것 같아요

taeyoungs replied:
확실히 아이템 객체를 들고 있는 편이 변경사항에 대응하거나 계산을 할 때 유리한 측면이 있는 것 같습니다.
체크된 아이템 객체 목록을 들고 있게 바꾸니 위 로직도 따로 id 값에 매핑되는 객체를 찾을 필요 없이 객체에 담긴 값을 더하기만 하면 되는 형태로 변경할 수 있었습니다.

```typescript
export const totalPriceOfCheckedCarts = (checkedCarts: Set<Cart>) =>
  Array.from(checkedCarts).reduce(
    (total, { product: { price }, count }) => total + price * count,
    0
  );
```

[242379c](https://github.com/next-step/react-shopping-cart/pull/23/commits/242379c3fe5898ea56fdb526e25f7c93d6033564)

jho2301 reviewed (changes requested):
안녕하세요 태영님!
> 평소에는 목록에서 체크 여부를 판단할 때 isChecked라는 속성을 따로 추가해주곤 했는데 이 방법 대신 Set 자료구조를 이용해서 체크된 항목들의 id를 관리하는 방식으로 진행해보았습니다.

하단 코멘트에 남겨보았듯이 체크된 카트객체를 인메모리에 들고 있어도 좋지 않을까 의견드려봅니다.
코멘트 자유롭게 답 부탁드릴게요.
감사합니다!

taeyoungs committed:
feat: 무한 스크롤을 위한 useInfiniteFetch 훅 추가

types/api.ts
- PaginationParams, BasePagination 타입 추가

hooks/useInfiniteFetch.ts
- page, hasNextPage를 내부에서 관리
- fethcMore 함수를 반환하여 추가적인 데이터를 불러올 때 해당 함수를
  호출
- 캐시 데이터가 존재하는 경우 캐시 데이터를 활용하도록 하고 캐시
  데이터의 마지막 BasePagination 객체 내 pageNumber로
  useInfiniteFetch 내 페이지 변수를 설정

taeyoungs committed:
feat: 페이지네이션 상품 목록 fetcher 및 Mock 함수 추가

zestlee1106 commented on client/src/__test__/store/cart.test.ts:
대박 Jest 도입도 성공하셨군요... 👍👍

taeyoungs replied:
> 대박 Jest 도입도 성공하셨군요... 👍👍

시도했다는 것에 의의를 두고 있습니다. 🫠 테스트 코드가 있긴 한데 구멍이 뚫려있는 느낌이라서요. 🤣

zestlee1106 commented on client/src/api/addOrder.ts:
제가 이러한 api request 모듈화(?) 에 익숙치 않아서요!

혹시 service 별로 구분하신 것이 아닌, 각 호출하는 api 마다 파일을 만들면 더 좋을까요?  
(구분하기에 좋다든지...)

아니면 취향 차이일까요???

예시))  
`api/order.ts` 파일에 `addOrder`, `deleteOrder` 함수 추가  
vs
`api/addOrder.ts`, `api/deleteOrder` 파일 각각 추가

taeyoungs replied:
> 제가 이러한 api request 모듈화(?) 에 익숙치 않아서요!
> 
> 혹시 service 별로 구분하신 것이 아닌, 각 호출하는 api 마다 파일을 만들면 더 좋을까요? (구분하기에 좋다든지...)
> 
> 아니면 취향 차이일까요???
> 
> 예시)) `api/order.ts` 파일에 `addOrder`, `deleteOrder` 함수 추가 vs `api/addOrder.ts`, `api/deleteOrder` 파일 각각 추가

저도 익숙하진 않습니다만 보통 api 폴더 하위에 대분류(ex. product, order)를 하나 두고 그 아래에 addOrder, deleteOrder 파일을 가각 추가하는 식으로 작성하긴 합니다(지극히 개인적으로 이 방식이 분류해놓은 게 한눈에 들어온달까요 ..? 🤣). 함수가 일정 개수 이하 이거나 파일을 열었을 때 한눈에 들어온다면 첫 번째 방식도 괜찮을 것 같네요. 🤔

taeyoungs committed:
feat: Intersection Oberserver 관련 컴포넌트 및 훅

taeyoungs committed:
mod: 캐시가 있을 경우 hasNextPage 업데이트 추가

taeyoungs committed:
mod: 상품 목록 Skeleton이 캐시를 이용하도록 변경

taeyoungs committed:
feat: Inifinite Fetch 및 Intersection Observer 추가

taeyoungs committed:
mod: 커스텀 훅 내 함수 메모이제이션 처리

taeyoungs committed:
mod: 객체가 동작의 주체가 될 수 있게 변경

zestlee1106 reviewed (commented):
순회 돌러 왔습니다.. 고생하셨습니다!! 

전체적으로 코드가 매우 깔끔해서 많이 배우고 갑니다... 👍👍

taeyoungs committed:
mod: useFetch 내 trigger를 shouldRefetch로 변경

taeyoungs committed:
feat: useMutation에 mutateAsync 함수 추가

taeyoungs committed:
chore: JSDoc으로 페이지 링크 진입 가능하게 변경

taeyoungs committed:
feat: apiClient 추가 및 이를 활용한 방식으로 변경

taeyoungs committed:
feat: 전역 에러 바운더리 추가

taeyoungs committed:
mod: 체크된 Cart의 id가 아닌 객체를 저장하도록 변경

taeyoungs commented:
안녕하세요 환오님!
피드백을 반영하는데 너무 오래 걸렸네요. 🥲
최근에 회사 일로 스트레스를 받기 시작하면서 많이 게을러진 것 같습니다. 🫠 🤦🏻‍♂️
어느새 또 변명을 .. 변명은 여기까지 하고 다시 본론으로 가겠습니다!

### 피드백 체크리스트

- [x] `src/api/deleteCarts.ts`의 5-14 라인을 사용하는 api 에 무관하게 추상화 해보기 ([adf0b28](https://github.com/next-step/react-shopping-cart/pull/23/commits/adf0b2828e1d581b83ff6170a78c705d074c59d3))
- [x] IDE에서 바로 페이지로 진입할 수 있게 JSDoc을 활용하기 ([a1a0ce0](https://github.com/next-step/react-shopping-cart/pull/23/commits/a1a0ce0bdc58100b7afcda1a23bfcf9a3380e87d))
- [x] useFetch 내 trigger를 shouldRefetch로 이름에서 역할을 구체화 시켜보기 ([bfcbf39](https://github.com/next-step/react-shopping-cart/pull/23/commits/bfcbf390af07fe37a8cec9e9e565ea9fce8b64f6))
- [x] useQuery의 mutateAsync와 같이 훅에 콜백을 전달하지 않고 프로미스 형태로 사용할 수 있게 만들어 동작을 보다 명확하게 파악할 수 있게 만들기 ([4df9a78](https://github.com/next-step/react-shopping-cart/pull/23/commits/4df9a784fac9ae5ff989636e14d29d844ce63619))
- [x] 구조 분해 할당하지 않고 객체가 동작의 주체가 될 수 있도록 만들어 보기 ([b4a140d](https://github.com/next-step/react-shopping-cart/pull/23/commits/b4a140de3a7f0f92d57fb35c26677995259d33b6))
- [x] 페이지가 미처 핸들링하지 못한 에러도 처리하는 전역 에러바운더리 추가하기 ([58330ab](https://github.com/next-step/react-shopping-cart/pull/23/commits/58330abdc1816ed87c6adbddfd275c1655b1ff93))
- [x] 커스텀훅은 어디서 사용될지에 대한 정보를 알지 못하기 때문에 최대한 메모이제이션 ([6d7e237](https://github.com/next-step/react-shopping-cart/pull/23/commits/6d7e237c705437618bbf9d0e04fc7acd35ba4d14))
- [x] 체크된 아이템 객체의 아이디 값만을 목록으로 가지고 있는 것이 아닌 아이템 객체 자체를 목록으로 관리하는 방식으로 변경하기 ([242379c](https://github.com/next-step/react-shopping-cart/pull/23/commits/242379c3fe5898ea56fdb526e25f7c93d6033564))

### 추가 구현사항

- [x] 상품목록 - 페이징 혹은 인피니티 스크롤 적용
  - [feat: 무한 스크롤을 위한 useInfiniteFetch 훅 추가](https://github.com/next-step/react-shopping-cart/pull/23/commits/124aa794a66d4db215f5231bd784c1d9138db322) 
  - [feat: 페이지네이션 상품 목록 fetcher 및 Mock 함수 추가](https://github.com/next-step/react-shopping-cart/pull/23/commits/1988f5e5fde5ea1ac9ff44b67da157cac71db829)
  - [feat: Intersection Oberserver 관련 컴포넌트 및 훅](https://github.com/next-step/react-shopping-cart/pull/23/commits/ae88e2a5455bf6ddad4e1bbd8312d18ad7824b2e)
  
인피니트 스크롤과 mutateAsync 쪽에서 생각보다 애를 많이 먹었네요.
시간 편하실 때 확인해주세요! 🙇‍♂️
시간 내서 리뷰해주셔서 항상 감사합니다!

jho2301 commented on client/src/hooks/useInfiniteFetch.ts:
```diff
@@ -0,0 +1,128 @@
+
+  const [state, dispatch] = useReducer(fetchReducer, initialState);
+  const [shouldRefetch, setShouldRefetch] = useState(false);
+  const [page, setPage] = useState(-1);
```
유효하지 않은 값을 나타내고자하셨다면 의미가 별도로 존재하는 매직값보다는 null이 있어도 괜찮을 것 같아요!

taeyoungs replied:
음 .. 확실히 -1이라서 더 헷갈릴 여지가 있는 것 같습니다.
환오님 말씀처럼 매직 넘버 보다는 null이 좋을 것 같아서 변경했습니다!

[c1ff2fd](https://github.com/next-step/react-shopping-cart/pull/23/commits/c1ff2fdb737611361796a66c58918589fd8d41f5)

jho2301 commented on client/src/api/addOrder.ts:
```diff
@@ -1,11 +1,10 @@
-    method: 'POST',
-    body: JSON.stringify(ids),
-  });
+  const response = await apiClient.post(API.ORDERS, { body: ids });
```
13번째 라인까지 fetch api의 관심사 인것 같아서 아래라인까지 포함한다면 더욱 좋지 않을까 생각해봅니다

taeyoungs replied:
고민하고 있던 부분이었는데 집어주셔서 바로 관심사 영역을 확장해 보았습니다!

[99cf26c](https://github.com/next-step/react-shopping-cart/pull/23/commits/99cf26c2e6ee8460f6c3c46f20cf62c7988a97e5)

jho2301 commented on client/src/api/addOrder.ts:
```diff
@@ -1,11 +1,10 @@
-    method: 'POST',
-    body: JSON.stringify(ids),
-  });
+  const response = await apiClient.post(API.ORDERS, { body: ids });
```
13번째 라인까지 fetch api의 관심사 인것 같아서 아래라인까지 포함한다면 더욱 좋지 않을까 생각해봅니다

jho2301 reviewed (approved):
반영사항 확인했습니다 태영님!
꼼꼼하게 반영해주셔서 감사합니다
전반적으로 이견있는 부분이 없어 머지하도록할게요 스텝3에서 뵙겠습니다!

taeyoungs committed:
mod: apiClient 관심사 영역 확장

taeyoungs committed:
mod: 유효하지 않은 값을 매직넘버 대신 null로 대체

taeyoungs commented:
안녕하세요 환오님!
리뷰 감사합니다. 🙇‍♂️

아직 머지가 되지 않은 김에 코멘트 남겨주신 부분 추가로 반영하고 한번 더 슬랙으로 요청드렸습니다!
